### DISCUSSION TITLE
/fp and the associated predefined macros

### DISCUSSION THREAD

**Discussion by kornman00** [NewWorldComingSoon/llvm-msvc-issues#33] [Q&A]
Near as I can tell, the MSVC driver for clang (clang-cl) does not define the predefined `_M_FP_*` macros based on what /fp setting is used

Consider these ClCompile msbuild settings
```xml
<FloatingPointModel>Precise</FloatingPointModel>
<FloatingPointExceptions>true</FloatingPointExceptions>
```

When I use VS2022's (17.4.4) ClangCL it will invoke it
`C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\Llvm\x64\bin\clang-cl.exe ... /fp:precise /fp:except ...`

However, it does not define the following from https://learn.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=msvc-170
`_M_FP_EXCEPT Defined as 1 if the /fp:except or /fp:strict compiler option is set. Otherwise, undefined.`
`_M_FP_PRECISE Defined as 1 if the /fp:precise compiler option is set. Otherwise, undefined.`

So e.g. this error will trigger
```C++
#ifndef _M_FP_PRECISE
#error no _M_FP_PRECISE 
#endif
```

More importantly, this ultimately leads to `_STD_VECTORIZE_WITH_FLOAT_CONTROL` being defined as `1`
```C++
// Controls whether the STL will force /fp:fast to enable vectorization of algorithms defined
// in the standard as special cases; such as reduce, transform_reduce, inclusive_scan, exclusive_scan
#ifndef _STD_VECTORIZE_WITH_FLOAT_CONTROL
#ifdef _M_FP_EXCEPT
#define _STD_VECTORIZE_WITH_FLOAT_CONTROL 0
#else // ^^^ floating-point exceptions enabled / floating-point exceptions disabled (default) vvv
#define _STD_VECTORIZE_WITH_FLOAT_CONTROL 1
#endif // _M_FP_EXCEPT
#endif // _STD_VECTORIZE_WITH_FLOAT_CONTROL
```

Which clang-cl then chokes on the `numeric` header
```
C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.34.31933\include\numeric(52,9): error : '#pragma float_control(precise, off)' is illegal when except is enabled
C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.34.31933\include\numeric(52,9): error : #pragma float_control(precise, off, push)
C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.34.31933\include\numeric(52,9): error :         ^
```

snippet from `MSVC\14.34.31933\include\numeric`
```C++
#if _HAS_CXX17
#if _STD_VECTORIZE_WITH_FLOAT_CONTROL
template <class _InIt, class _Ty, class _BinOp>
inline constexpr bool _Plus_on_arithmetic_ranges_reduction_v =
    conjunction_v<is_arithmetic<_Ty>, is_arithmetic<remove_pointer_t<_InIt>>, is_same<plus<>, _BinOp>>;

#pragma float_control(precise, off, push)
template <class _InIt, class _Ty>
_Ty _Reduce_plus_arithmetic_ranges(_InIt _First, const _InIt _Last, _Ty _Val) {
    // return reduction, plus arithmetic on contiguous ranges case
#pragma loop(ivdep)
    for (; _First != _Last; ++_First) {
        _Val += *_First;
    }

    return _Val;
}
#pragma float_control(pop)

#else // ^^^ _STD_VECTORIZE_WITH_FLOAT_CONTROL ^^^ // vvv !_STD_VECTORIZE_WITH_FLOAT_CONTROL vvv
template <class _InIt, class _Ty, class _BinOp>
inline constexpr bool _Plus_on_arithmetic_ranges_reduction_v = false;
#endif // _STD_VECTORIZE_WITH_FLOAT_CONTROL
```

I stumbled upon this repo while searching for answers related to other issues with MSVC compatibility, so figured I would flag this issue here at least. Please let me know if I should report this directly to LLVM instead. Thank you!

**Comment by gmh5225** [comment]
Can you give me a code example? 
I was unable to reproduce the problem you described in my own tests.


Command
```
clang-cl.exe /c /Z7 /nologo /W3 /WX- /diagnostics:column /Od /D PSAPI_VERSION=1 /D _DEBUG /D _CONSOLE /D _UNICODE /D UNICODE /EHsc /MDd /GS /fp:precise /fp:except /permissive- /Fox64\Debug\ /external:W3 /Gd /TP --target=x86_64-pc-windows-msvc Source.cpp
```
Source.cpp
```C++
#include <Windows.h>
#include <stdio.h>


int main()
{
	auto x = _M_FP_PRECISE;
	printf("%d\n", x);
	return 0;
}
```
![image](https://user-images.githubusercontent.com/13917777/213073619-75404748-3afc-43f4-82fb-2f46a55c3d9e.png)

You can install the llvm-msvc to try the effect
https://github.com/NewWorldComingSoon/llvm-msvc-build

**Comment by kornman00** [comment]
Thanks for the reply, I'm guessing previous changes may have already addressed this macro? I think 17.4.4 bundles with 15.0.1 but when I searched 15.0.4 or whichever is the latest LLVM source I didn't see any references to permutations of the macro (edit: never mind, appears vcpkg actually pulled down 14.0.4, sorry). Github doesn't support searching forked repos so I wasn't able to search for instances before making this new issue (and when I glanced over the already closed issues I didn't see anything that already reported it).

I'll download the forked build later this week and try it directly :)

I see LLVM added the /fp aliases back in 2015, with a note "It probably doesn't match cl.exe's behaviour
completely" heh https://github.com/llvm/llvm-project/commit/7a2e5c260c89fe40f0d65e165a83d75a34bc587e

For my own edification, do you have a sense of how often the LLVM project takes your MSVC compatibility changes? Or is the project meant to be a maintained fork of LLVM? Definitely appreciate the work either way.

**Comment by gmh5225** [comment]
If you want to use the official llvm. I recommend you to use version (15.0.7/16).

`llvm-msvc` is a project that forked LLVM focused on MSVC Compatibility.
Because there are more hacky operations, a lot of code can not be submitted directly to the official. So there is this branch.

`llvm-msvc` will regularly(1/7 days) merge the official llvm.

**Comment by kornman00** [comment]
Sorry for taking so long to circle back to this: I can confirm that this is not an issue with llvm-msvc (using whatever build was available on the 25th). Cheers!
